File: articles/nonlinearity-feedback-biological-regulation/python/negative_feedback.py
"""
Negative feedback homeostasis scenarios.

Run:
    python python/negative_feedback.py
"""

from pathlib import Path

import pandas as pd

from nonlinear_feedback_core import NegativeFeedbackParameters, simulate_negative_feedback


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "negative_feedback_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)
    rows = []

    for _, row in scenarios.iterrows():
        trajectory = simulate_negative_feedback(
            NegativeFeedbackParameters(
                x0=row["x0"],
                set_point=row["set_point"],
                k=row["k"],
                dt=row["dt"],
                t_end=row["t_end"],
            )
        )

        rows.append(
            {
                "scenario": row["scenario"],
                "initial_state": trajectory["state"].iloc[0],
                "final_state": trajectory["state"].iloc[-1],
                "set_point": row["set_point"],
                "final_error": abs(trajectory["state"].iloc[-1] - row["set_point"]),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/nonlinearity-feedback-biological-regulation/r/negative_feedback.R
# Negative feedback homeostasis workflow.

simulate_negative_feedback <- function(x0, set_point, k, dt = 0.05, t_end = 30) {
  time <- seq(0, t_end, by = dt)
  x <- numeric(length(time))
  x[1] <- x0

  for (i in 2:length(time)) {
    dx <- -k * (x[i - 1] - set_point)
    x[i] <- x[i - 1] + dx * dt
  }

  data.frame(time = time, state = x)
}

trajectory <- simulate_negative_feedback(180, 100, 0.18)

summary_df <- data.frame(
  initial_state = trajectory$state[1],
  final_state = tail(trajectory$state, 1),
  set_point = 100,
  final_error = abs(tail(trajectory$state, 1) - 100)
)

print(round(summary_df, 5))
